### DISCUSSION TITLE
Text Clipping

### DISCUSSION THREAD

**Discussion by JamesPScottPharm3r** [QuestPDF/QuestPDF#789] [Q&A] [answered]
Hello Marcin and community, I've been trying to figure out how to clip text instead of wrapping within a table layout.

If a certain piece of text was longer than the width of the Canvas available in the IContainer that I would remove some characters at the end of the string and replace with "..." so everything would fit on one line.  

(E.g.  "This really long piece of text I don't want to wrap" could become "This really long piece...")

So far I have been trying to use SKFont.MeasureText and Canvas.DrawText, but neither seem to be responding correctly to font sizes etc...  I'll put more info into this as I continue my investigation, but thought the community may have a quicker path to a solution.

Thanks all!

James

**Comment by JamesPScottPharm3r** [comment]
Hello community, checking in with what may be a potential solution.  It seems to work ok...  In order for it to work well your container must have a specific width and height.  Otherwise I found the drawing regions were wrong and text would be rendered in spots you do not want.

```csharp
container.Canvas((canvas, size) =>
        {
            using var paint = new SkiaSharp.SKPaint
            {
                IsAntialias = true,
                Typeface = _font.Typeface,
                TextSize = fontSize
            };

            SKRect textDimensions = new SKRect();
            
            // I was using SKFont.MeasureText with a paint object before but it was NOT working at all.  This seems
            // to work much better.  
            paint.MeasureText(text, ref textDimensions);
            //_font!.MeasureText(testspanushort, out textDimensions, paint);
            
            //Console.WriteLine($"MeasureText: Font size: {paint.TextSize} - Text: {text} - Dimensions - {textDimensions.Width}x{textDimensions.Height}");

            if (textDimensions.Width > size.Width)
            {
                // lets find the difference between the width we have and the width we can draw in
                float longerThanCanvas = textDimensions.Width - size.Width;
                // from that we will get the approximate amount of characters we need to drop
                int numberOfCharsToDrop = (int)Math.Round(longerThanCanvas / (textDimensions.Width / text.Length)) + 3;
                // because I'm using an average above I thought it best to always do one less char...
                numberOfCharsToDrop++;

                text = text.Substring(0, text.Length - numberOfCharsToDrop) + "...";
            }
            
            // not sure why, but adding the fontSize to 0 seems to put the text in the correct spot.. maybe its because
            // its start drawing text in lower left hand corner?  so if we specify 0 it actually draws into negatives?
            canvas.DrawText(text, 0, 0 + fontSize, paint);
        });
```

**Comment by MarcinZiabek** [comment]
This feature should be available soon as part of this effort: https://github.com/QuestPDF/QuestPDF/discussions/622. It is partially implemented but not available yet. 😄

**Comment by JamesPScottPharm3r** [comment]
That's great news Marcin, thank you.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

The functionality you described is now available through the [Clamp Line with Ellipsis](https://www.questpdf.com/api-reference/text/paragraph-style.html#clamp-line-with-ellipsis) paragraph style.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
